### DISCUSSION TITLE
New Langauge Finetuning

### DISCUSSION THREAD

**Discussion by TarekWSw** [QwenLM/Qwen3-TTS#208] [Q&A]
This is training script i am using and those are the configs, is this combination should yield are good results, or i should extend the training epochs? I am fine-tuning for the arabic langauge. My dataset is around 103h (33k samples of audio segments, ranging between 3-16s)
Also is it better to modify the loss in `qwen_tts/core/models/modeling_qwen3_tts.py, line 1243:` as mentioned here https://github.com/QwenLM/Qwen3-TTS/issues/179#issuecomment-3870059313 ?

```
  batch_size: 4
  learning_rate: 2e-6
  weight_decay: 0.01
  num_epochs: 5
  gradient_accumulation_steps: 8
  mixed_precision: "bf16"
  max_grad_norm: 1.0
  sub_talker_loss_weight: 0.3
```

```
import ....

target_speaker_embedding = None

def train(cfg: SFTConfig) -> None:
    """Run supervised fine-tuning on a Qwen3-TTS base model."""
    global target_speaker_embedding

    accelerator = Accelerator(
        gradient_accumulation_steps=cfg.training.gradient_accumulation_steps,
        mixed_precision=cfg.training.mixed_precision,
        log_with="tensorboard",
        project_dir=cfg.output.logs_dir,
    )

    qwen3tts = Qwen3TTSModel.from_pretrained(
        cfg.model.init_model_path,
        torch_dtype=torch.bfloat16,
        attn_implementation=cfg.model.attn_implementation,
    )
    config = AutoConfig.from_pretrained(cfg.model.init_model_path)

    with open(cfg.data.train_jsonl) as f:
        train_data = [json.loads(line) for line in f]

    dataset = TTSDataset(
        train_data, qwen3tts.processor, config, jsonl_path=cfg.data.train_jsonl
    )
    train_dataloader = DataLoader(
        dataset,
        batch_size=cfg.training.batch_size,
        shuffle=True,
        collate_fn=dataset.collate_fn,
    )

    optimizer = AdamW(
        qwen3tts.model.parameters(),
        lr=cfg.training.learning_rate,
        weight_decay=cfg.training.weight_decay,
    )

    model, optimizer, train_dataloader = accelerator.prepare(
        qwen3tts.model, optimizer, train_dataloader
    )

    model.train()

    for epoch in range(cfg.training.num_epochs):
        for step, batch in enumerate(train_dataloader):
            with accelerator.accumulate(model):
                input_ids = batch["input_ids"]
                codec_ids = batch["codec_ids"]
                ref_mels = batch["ref_mels"]
                text_embedding_mask = batch["text_embedding_mask"]
                codec_embedding_mask = batch["codec_embedding_mask"]
                attention_mask = batch["attention_mask"]
                codec_0_labels = batch["codec_0_labels"]
                codec_mask = batch["codec_mask"]

                speaker_embedding = model.speaker_encoder(
                    ref_mels.to(model.device).to(model.dtype)
                ).detach()
                if target_speaker_embedding is None:
                    target_speaker_embedding = speaker_embedding

                input_text_ids = input_ids[:, :, 0]
                input_codec_ids = input_ids[:, :, 1]

                input_text_embedding = (
                    model.talker.model.text_embedding(input_text_ids)
                    * text_embedding_mask
                )
                input_codec_embedding = (
                    model.talker.model.codec_embedding(input_codec_ids)
                    * codec_embedding_mask
                )
                input_codec_embedding[:, 6, :] = speaker_embedding

                input_embeddings = input_text_embedding + input_codec_embedding

                outputs = model.talker(
                    inputs_embeds=input_embeddings,
                    attention_mask=attention_mask,
                    labels=codec_0_labels,
                    output_hidden_states=True,
                )

                hidden_states = outputs.hidden_states[0][-1]
                target_codec_mask = codec_mask[:, 1:]
                talker_hidden_states = hidden_states[:, :-1][target_codec_mask]
                talker_codec_ids = codec_ids[:, 1:][target_codec_mask]

                _, sub_talker_loss = model.talker.forward_sub_talker_finetune(
                    talker_codec_ids, talker_hidden_states
                )

                loss = (
                    outputs.loss + cfg.training.sub_talker_loss_weight * sub_talker_loss
                )

                accelerator.backward(loss)

                if accelerator.sync_gradients:
                    accelerator.clip_grad_norm_(
                        model.parameters(), cfg.training.max_grad_norm
                    )

                optimizer.step()
                optimizer.zero_grad()

            if step % 10 == 0:
                accelerator.print(
                    f"Epoch {epoch} | Step {step} | Loss: {loss.item():.4f}"
                )

        if accelerator.is_main_process:
            output_dir = os.path.join(
                cfg.output.model_path, f"checkpoint-epoch-{epoch}"
            )
            shutil.copytree(cfg.model.init_model_path, output_dir, dirs_exist_ok=True)

            input_config_file = os.path.join(cfg.model.init_model_path, "config.json")
            output_config_file = os.path.join(output_dir, "config.json")
            with open(input_config_file, encoding="utf-8") as f:
                config_dict = json.load(f)
            config_dict["tts_model_type"] = "custom_voice"
            talker_config = config_dict.get("talker_config", {})
            talker_config["spk_id"] = {cfg.speaker.name: 3000}
            talker_config["spk_is_dialect"] = {cfg.speaker.name: False}
            config_dict["talker_config"] = talker_config

            with open(output_config_file, "w", encoding="utf-8") as f:
                json.dump(config_dict, f, indent=2, ensure_ascii=False)

            unwrapped_model = accelerator.unwrap_model(model)
            state_dict = {
                k: v.detach().to("cpu") for k, v in unwrapped_model.state_dict().items()
            }

            drop_prefix = "speaker_encoder"
            keys_to_drop = [k for k in state_dict if k.startswith(drop_prefix)]
            for k in keys_to_drop:
                del state_dict[k]

            weight = state_dict["talker.model.codec_embedding.weight"]
            state_dict["talker.model.codec_embedding.weight"][3000] = (
                target_speaker_embedding[0].detach().to(weight.device).to(weight.dtype)
            )
            save_path = os.path.join(output_dir, "model.safetensors")
            save_file(state_dict, save_path)

```
